src/exec/use_pty/pipe.rs

```diff
@@ -219,16 +271,169 @@ impl<R: Read, W: Write> Buffer<R, W> {
 
     /// Flush this buffer, ensuring that all the contents of its internal buffer are written.
     fn flush(&mut self, write: &mut W) -> io::Result<()> {
-        // This is the busy section of the buffer.
-        let buffer = &self.internal[self.start..self.end];
+        // Remove bytes from the buffer and write them.
+        self.internal.remove(write)?;
 
-        // Write the complete busy section to `write`.
-        write.write_all(buffer)?;
+        write.flush()
+    }
+}
 
-        // If we were able to write all the busy section, we can mark the whole buffer as free.
-        self.start = 0;
-        self.end = 0;
+#[cfg(test)]
+mod tests {
+    use super::RingBuffer;
 
-        write.flush()
+    #[test]
+    fn empty_buffer_is_empty() {
+        let buf = RingBuffer::new();
+
+        assert!(buf.is_empty());
+    }
+
+    #[test]
+    fn full_buffer_is_full() {
+        let mut buf = RingBuffer::new();
+
+        let inserted_len = buf.insert(&mut [0x45; RingBuffer::LEN].as_slice()).unwrap();
+        assert_eq!(inserted_len, RingBuffer::LEN);
+
+        assert!(buf.is_full());
+    }
+
+    #[test]
+    fn buffer_is_fifo() {
+        let mut buf = RingBuffer::new();
+
+        let expected = (0..=u8::MAX).collect::<Vec<u8>>();
+        let inserted_len = buf.insert(&mut expected.as_slice()).unwrap();
+        assert_eq!(inserted_len, expected.len());
+
+        let mut found = vec![];
+        let removed_len = buf.remove(&mut found).unwrap();
+        assert_eq!(removed_len, expected.len());
+
+        assert_eq!(expected, found);
+    }
+
+    #[test]
+    fn insert_into_empty_buffer_with_offset() {
+        const HALF_LEN: usize = RingBuffer::LEN / 2;
+        let mut buf = RingBuffer::new();
+
+        // This should leave the buffer empty but with the start field pointing to the middle of
+        // the buffer.
+        // ┌───────────────────┐
+        // │                   │
+        // └───────────────────┘
+        //           ▲
+        //           │
+        //         start
+        buf.insert(&mut [0u8; HALF_LEN].as_slice()).unwrap();
```

Comment:
in this particular case you could set the initial condition by directly setting the value of `start` to `HALF_LEN`. that way this unit test does not rely on `insert` and `remove` correctly working in cases 1.1 and 2.1.

Reply:
Yeah I thought about it but I think I'd rather leave those tests as implementation agnostic as possible.